### PR TITLE
Improve Merge and Commit logic

### PR DESCRIPTION
**Pull request by Stummi**
### Type of Change

- [x] Bugfix
- [x] Enhancement / new feature
- [x] Refactoring
- [ ] Documentation

### Description

- Create reproducible commit hashes by using a fixed date/time for the merge commit
- Remove optimistic option (It is never actually used, and adds unnecessary complexity)
- Preserve parent refs for dev merge commit
- Replace hasDiff logic with a comparison on commit hashes

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [x] Review the [Contributing Guideline](https://github.com/Staffbase/yamllint-action/blob/main/CONTRIBUTING.md) and sign CLA
- [ ] Reference relevant issue(s) and close them after merging

### CODE REVIEW

**Comment by Stummi** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/autodev.ts [hunk 1]
```diff
@@ -122,12 +123,18 @@ const merge = async (
...
   const success: Pull[] = []
   for (const pull of pulls) {
     try {
-      await exec(`git merge origin/${pull.branch}`)
+      await exec(`git merge origin/${pull.branch}`, undefined, {
+        env: {
+          GIT_COMMITTER_DATE: commitDate,
+          GIT_AUTHOR_DATE: commitDate
+        }
+      })
```
[the remark this answers is not in the document]
Comment (Stummi) [reply src/autodev.ts:137]: ah, yes, you are right. Is there a particular reason why it is done this way? Why not just keep a number of merge commits?

[the remark this answers is not in the document]
Comment (staffbase-robert) [reply src/autodev.ts]: A single commit with X merged branches is much easier to read and understand compared to X merge commits + commits that have been made in the branches (we're not using squash merge). For some large repos you have to read trough 2 pages of commits to understand what has been merged into dev.

[the remark this answers is not in the document]
Comment (staffbase-robert) [reply src/autodev.ts]: If it's the goal to create the same commit hash, when the same branches have been merged, then it should be enough to just change the commit + author dates of the last commit. [source](https://blog.thoughtram.io/git/2014/11/18/the-anatomy-of-a-git-commit.html)

File: src/autodev.ts [hunk 2]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
   await exec('git fetch')
   await exec(`git config --global user.email "${email}"`)
   await exec(`git config --global user.name "${user}"`)
-  await exec(`git checkout ${branch}`)
-  await exec(`git reset --hard origin/${base}`)
+  const commitDate = await execAndSlurp(
+    `git show -s --format='%ci' origin/${base}`
+  )
+
+  await exec(`git checkout ${base}`)
```
Comment (Stummi) [inline src/autodev.ts:102]: Creation of new branch was moved to the end of logic. Its not yet needed, just start on the base branch

File: src/autodev.ts [hunk 3]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
...
-    if (await hasDiff('HEAD', `origin/${branch}`)) {
-      await exec('git push -f')
-      info(
-        `🎉 No Pull Requests found. Pushed changes, because "${branch}" and "${base}" diverged.`
-      )
-    } else {
-      info('🎉 No Pull Requests found. Nothing to merge.')
-    }
-    return
+    info('🎉 No Pull Requests found. Nothing to merge.')
```
Comment (Stummi) [inline src/autodev.ts:105]: There is no need anymore to duplicate the check for a diff here

File: src/autodev.ts [hunk 4]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
...
+      updateLabel,
+      commitDate
+    )
+    info(message)
   }
-
-  const message = optimistic
-    ? await merge(base, pulls, updateComment, updateLabel)
-    : await mergeAll(pulls, updateComment, updateLabel)
+  await exec(`git checkout -B ${branch}`)
```
Comment (Stummi) [inline src/autodev.ts:116]: Create the new dev branch on the current commit (which is either `$base` if no dev pulls exists, or the newly created merge commit)

File: src/autodev.ts [hunk 5]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
...
 
   // only push to defined branch if there are changes
+  await exec('git fetch')
```
Comment (Stummi) [inline src/autodev.ts:119]: fetch again. This reduces (though not completely eliminates) a race condition where another parallel invocation of the autodev action could result in having a inconsistent origin/dev ref

File: src/autodev.ts [hunk 6]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
...
   if (await hasDiff('HEAD', `origin/${branch}`)) {
-    await exec('git push -f')
+    // ignore any errors
+    await exec('git push -f', undefined, {
+      ignoreReturnCode: true
```
Comment (Stummi) [inline src/autodev.ts:123]: Command will still fail if run into the (earlier mentioned) race condition here. But in this case we can just ignore the failure (basically it means another invocation of the dev workflow beat us in pushing the same state)

File: src/autodev.ts [hunk 7]
```diff
@@ -75,44 +95,41 @@ const autoDev = async (): Promise<void> => {
...
-  await exec(`git diff ${a}..${b}`, undefined, {
-    listeners: {
-      stdout: () => {
-        ret = true
-      }
-    }
-  })
-
-  return ret
+  return (
```
Comment (Stummi) [inline src/autodev.ts:129]: Since we produce reproducible commit hashes now, just compare them to see if the dev state changed

**Review by staffbase-robert — COMMENTED** [review]
<summary>The action fails now with this message:
<details>

```bash
 /usr/bin/git fetch
/usr/bin/git config --global user.email staffbot@staffbase.com
/usr/bin/git config --global user.name autodev-action
/usr/bin/git show -s --format='%ci' origin/main
'2022-06-03 13:55:05 +0200'
/usr/bin/git checkout main
Switched to a new branch 'main'
branch 'main' set up to track 'origin/main'.
/usr/bin/git merge origin/feature
Updating 94e3cc6..767c0a4
Fast-forward
 README.md | 2 ++
 1 file changed, 2 insertions(+)
/usr/bin/git reset origin/main
Unstaged changes after reset:
M	README.md
/usr/bin/git add -A
/usr/bin/git commit -m AutoDev Merge

The following branches have been merged:
- feature
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: empty ident name (for <runner@fv-az39-531.yumyzuo2tp0ejekjjwodrrku4b.bx.internal.cloudapp.net>) not allowed
```
</details>
</summary>

**Review by flaxel — COMMENTED** [review]
Great improvement! 🎉

**Comment by flaxel** [comment]
> The action fails now with this message

You can use [this example PR](https://github.com/flaxel/test-gha/pull/35) to reproduce the issue.

As a first check can you can add the command to check the configuration:

```bash
git config --global --list
```

Otherwise I think we could add the [information as environment variables](https://git-scm.com/docs/git-commit#_commit_information). Or we could use [the option](https://stackoverflow.com/a/51612036/10951752) `--reset-author`.

File: src/autodev.ts (hunk 3 above)
Comment (Stummi) [reply src/autodev.ts:105]: No, I would still go to the push, so when all pulls get the "dev" label removed, the autodev action will revert the dev branch to main

**Comment by Stummi** [comment]
I am not sure what the exact reason is for the original issue, however it seems to work when I set the git config locally.

File: src/autodev.ts [hunk 8]
```diff
@@ -7,8 +7,30 @@ import {
   updateLabels
 } from './utils'
 import {getInput, info, setFailed} from '@actions/core'
+import {ExecOptions} from '@actions/exec/lib/interfaces'
 import {exec} from '@actions/exec'
 
+const execAndSlurp = async (
```
Comment (flaxel) [inline src/autodev.ts:13]: Could you add a little comment to this method?

**Comment by flaxel** [comment]
Looks really good, but my local changes differ from the `dist` folder. Could you run the command again?

```bash
npm run all
```

**Comment by Stummi** [comment]
I resolved the issue now

File: src/autodev.ts [hunk 9]
```diff
@@ -139,32 +160,28 @@ const merge = async (
...
+
+  await exec('git commit -m', [message], overrideDate)
+  // replace with graft commit so we can preserve commit parents
+  await exec(
+    `git replace --graft HEAD origin/${base}`,
+    success.map(p => `origin/${p.branch}`),
+    overrideDate
+  )
+  const rev = await execAndSlurp('git rev-parse HEAD')
+  await exec(`git checkout replace/${rev}`)
```
Comment (flaxel) [inline src/autodev.ts:181]: I guess this is normal and we can't prevent it? 

> You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.
>
> If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:
>
>  git switch -c <new-branch-name>
>
>Or undo this operation with:
>
>  git switch -
>
>Turn off this advice by setting config variable advice.detachedHead to false

Comment (flaxel) [reply src/autodev.ts:181]: And I don't know what I do differently, but when I run `npm run all`, I have differences again?!

Comment (Stummi) [reply src/autodev.ts:181]: > I guess this is normal and we can't prevent it?

You mean the message? I don't think so

Comment (Stummi) [reply src/autodev.ts:181]: > And I don't know what I do differently, but when I run `npm run all`, I have differences again?!

I guess the result of `npm run all` is not exactly reproducible,

Comment (Stummi) [reply src/autodev.ts:181]: I did  a rebase now, removed `yarn.lock`, ran `yarn` again and then reran `npm run all`. Hope this result is now reproducible on your end

**Review by flaxel — APPROVED** [review]

**Review by staffbase-robert — APPROVED** [review]
It works for me now. 
The action now adds all commits of all merged pull requests to the commit history on the dev-branch. It doesn't hurt, but is this really necessary? I rarely check this commit history and only check out the "squash" commit at the end to understand which branches have been combined. 

A "squash" option would be cool :)

**Pull request merged**